pkg/catalogv2/git/config.go

```diff
@@ -0,0 +1,196 @@
+package git
+
+import (
+	"fmt"
+	"net/url"
+
+	gogit "github.com/go-git/go-git/v5"
+	transport "github.com/go-git/go-git/v5/plumbing/transport"
+	"github.com/rancher/rancher/pkg/settings"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// extendedRepo embeds repository interface defining git-related operations.
+// This is needed in order to generate mocks for unit-tests
+type extendedRepo struct {
+	repository
+}
+
+// repoOperation holds the config of a git repository and the repo instance.
+type repoOperation struct {
+	config    *git
+	localRepo *gogit.Repository
+	auth      transport.AuthMethod // interface
+	cloneOpts *gogit.CloneOptions
+	fetchOpts *gogit.FetchOptions
+}
+
+type Options struct {
+	Credential        *corev1.Secret
+	CABundle          []byte
+	InsecureTLSVerify bool
+	Headers           map[string]string
+}
+
+type git struct {
+	URL               string
+	Directory         string
+	username          string
+	password          string
+	agent             *agent.Agent
+	caBundle          []byte
+	insecureTLSVerify bool
+	secret            *corev1.Secret
+	headers           map[string]string
+	knownHosts        []byte
+}
+
+// gitForRepo constructs and returns a new git object for the given repository.
+// It requires a secret for authentication, the namespace, name of the repository,
+// the gitURL, a flag indicating if TLS verification should be skipped, and a CA bundle for SSL.
+// If the Git URL uses the SSH protocol, it checks if the URL is a valid SSH URL.
+// If the Git URL uses HTTP or HTTPS, it parses and verifies the URL.
+// It then constructs a directory path for the git repository.
+// It also sets an X-Install-Uuid header if an installation UUID is available.
+// If a CA bundle is provided, it converts the CA bundle from DER to PEM format, since Git requires PEM format.
+// In this case, insecureSkipTLS is set to false since a CA bundle is provided for secure communication.
+// Finally, it returns a new git object configured with these settings,
+// or an error if any step in this process fails.
+func gitForRepo(secret *corev1.Secret, namespace, name, gitURL string, insecureSkipTLS bool, caBundle []byte) (*extendedRepo, error) {
+	isGitSSH, err := isGitSSH(gitURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to verify the type of URL %s: %w", gitURL, err)
+	}
+	if !isGitSSH {
+		u, err := url.Parse(gitURL)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse URL %s: %w", gitURL, err)
+		}
+		if u.Scheme != "http" && u.Scheme != "https" {
+			return nil, fmt.Errorf("invalid git URL scheme %s, only http(s) and git supported", u.Scheme)
```

Comment:
The error message calls out that the `git` scheme (e.g. `git://`) is supported, but the code above says otherwise.